tempesta_fw/http_frame.c

```diff
@@ -0,0 +1,950 @@
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "lib/fsm.h"
+#include "lib/str.h"
+#include "procfs.h"
+#include "http.h"
+#include "http_frame.h"
+
+
+#define FRAME_CLI_MAGIC		"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"
+#define FRAME_CLI_MAGIC_LEN	24
+#define FRAME_SRVC1_SIZE	4
+#define FRAME_PRIORITY_SIZE	5
+#define FRAME_STNGS_ENTRY_SIZE	6
+#define FRAME_SRVC2_SIZE	8
+
+#define FRAME_STREAM_ID_MASK	((1U << 31) - 1)
+
+/* HTTP/2 frame types (RFC 7540 section 6).*/
+typedef enum {
+	HTTP2_DATA		= 0,
+	HTTP2_HEADERS,
+	HTTP2_PRIORITY,
+	HTTP2_RST_STREAM,
+	HTTP2_SETTINGS,
+	HTTP2_PUSH_PROMISE,
+	HTTP2_PING,
+	HTTP2_GOAWAY,
+	HTTP2_WINDOW_UPDATE,
+	HTTP2_CONTINUATION
+} TfwFrameType;
+
+/*
+ * HTTP/2 error codes (RFC 7540 section 7). Used in RST_STREAM
+ * and GOAWAY frames to report the reasons of the stream or
+ * connection error.
+ */
+#define FRAME_ECODE_PROTO	0x1
+#define FRAME_ECODE_SIZE_ERROR	0x6
+
+/*
+ * HTTP/2 frame flags. Can be specified in frame's header and
+ * are specific to the particular frame types (RFC 7540 section
+ * 4.1 and section 6).
+ */
+typedef enum {
+	HTTP2_F_ACK		= 0x01,
+	HTTP2_F_END_STREAM	= 0x01,
+	HTTP2_F_END_HEADERS	= 0x04,
+	HTTP2_F_PADDED		= 0x08,
+	HTTP2_F_PRIORITY	= 0x20
+} TfwFrameFlag;
+
+#define __FRAME_FSM_EXIT()						\
+do {									\
+	ctx->rlen = 0;							\
+	T_FSM_EXIT();							\
+} while (0)
+
+#define FRAME_FSM_EXIT(ret)						\
+do {									\
+	r = ret;							\
+	__FRAME_FSM_EXIT();						\
+} while (0)
+
+#define FRAME_FSM_FINISH()						\
+	T_FSM_FINISH(r, ctx->state);					\
+	*read += p - buf;
+
+#define FRAME_FSM_MOVE(st)						\
+do {									\
+	WARN_ON_ONCE(p - buf > len);					\
+	ctx->rlen = 0;							\
+	T_FSM_MOVE(st,							\
+		   if (unlikely(p - buf >= len)) {			\
+			   __fsm_const_state = st;			\
+			   T_FSM_EXIT();				\
+		   });							\
+} while (0)
+
+#define FRAME_FSM_NEXT()						\
+do {									\
+	WARN_ON_ONCE(p - buf > len);					\
+	ctx->rlen = 0;							\
+	if (unlikely(p - buf >= len)) {					\
+		__fsm_const_state = ctx->state;				\
+		T_FSM_EXIT();						\
+	}								\
+	T_FSM_NEXT();							\
+} while (0)
+
+#define FRAME_FSM_READ_LAMBDA(to_read, lambda)				\
+do {									\
+	WARN_ON_ONCE(ctx->rlen >= (to_read));				\
+	n = min_t(int, (to_read) - ctx->rlen, buf + len - p);		\
+	lambda;								\
+	p += n;								\
+	ctx->rlen += n;							\
+	if (unlikely(ctx->rlen < (to_read)))				\
+		T_FSM_EXIT();						\
```

Comment:
In this case we'll exit only if amount of already read bytes is less then needed amount, so we must continue reading in current state until `ctx->rlen == to_read`.

Reply:
Let's assume we received data chunk that has initial magic string, then SETTINGS frame header and part of the first setting in the SETTING frame.
We'll start at `HTTP2_RECV_CLI_START_SEQ`, read the magic string, then move to `HTTP2_RECV_FIRST_SETTINGS`, parse frame header, then move to `HTTP2_RECV_FRAME_SETTINGS`, where `FRAME_FSM_READ_SRVC()` will invoke `FRAME_FSM_READ_LAMBDA()` where we read part of a setting when data chunk ends. At that point we'll get here, call `T_FSM_EXIT()`, which jumps to `__fsm_done` defined in `T_FSM_FINISH()`, where `ctx->state` will be set to `__fsm_const_state`. Most probably it will be `0`, and not `HTTP2_RECV_FRAME_SETTINGS`, as expected. So when the next data chunk arrives, we'll start reading the magic string again.

Reply:
While we are jumping from state to state - `__fsm_const_state` is updated in `T_FSM_STATE`, so it should have the actual value in `T_FSM_FINISH`.

Reply:
Ah, I see now. Yes, that's OK.